File: Recrussion/php/fibonacci.php
<?php

declare(strict_types=1);

function fibonacci(int $n, array &$memo = []): int
{
    if ($n < 0) {
        throw new InvalidArgumentException('n must be non-negative');
    }

    if ($n === 0 || $n === 1) {
        return $n;
    }

    if (isset($memo[$n])) {
        return $memo[$n];
    }

    $memo[$n] = fibonacci($n - 1, $memo) + fibonacci($n - 2, $memo);
    return $memo[$n];
}

File: Recrussion/java/fibonacci.java
import java.util.HashMap;
import java.util.Map;

class Fibonacci {
  private static final Map<Integer, Long> memo = new HashMap<>();

  static {
    memo.put(0, 0L);
    memo.put(1, 1L);
  }

  public static long fibonacci(int n) {
    if (n < 0) {
      throw new IllegalArgumentException("n must be non-negative");
    }
    if (memo.containsKey(n)) {
      return memo.get(n);
    }
    long value = fibonacci(n - 1) + fibonacci(n - 2);
    memo.put(n, value);
    return value;
  }

  public static void run() {
    // intentionally empty
  }
}
